### DISCUSSION TITLE
Support for DNS names alongside IP addresses

### DISCUSSION THREAD

**Discussion by fencepost-error** [BlackZork/mqmgateway#46] [Ideas]
Would it be possible to support DNS names alongside IP addresses for devices?  At the moment the code passes the device address directly to modbus_new_tcp() which requires an IP address but this doesn't work for devices that use DHCP, to be able to use DNS names as well the code could check for !isdigit(address[0]) and then do something like:
```
addrinfo hints = {0}, *addr_info;
const char *ipAddr;

hints.ai_family = AF_INET; // Not AF_UNSPEC since libmodbus requires IPv4
hints.ai_socktype = SOCK_STREAM;
hints.ai_protocol = IPPROTO_TCP;
if(getaddrinfo(config.mAddress.c_str(), NULL, &hints, &addr_info) != 0)
  return false;
inetAddr = inet_ntoa(((sockaddr_in *) addr_info->ai_addr)->sin_addr);
```

**Comment by BlackZork** [comment]
Good idea, thanks! I need an async resolver for it.

**Reply by fencepost-error** [reply]
Do you really need an async resolver?  Sounds like a lot of complexity and, although I've only glanced through the code, I'm not sure why you'd need it, presumably the lookup is done once on start and if it fails you bail out.  It's of value in browsers but seems unnecessary here.

**Reply by BlackZork** [reply]
Lookup needs to be done in each modbus thread. I think it should be restarted if reconnect delay gets to 30s (it is incremented by 5s each try until 30s is reached), in case when IP address is changed in DHCP server.

In the best case it should be possible to SIGTERM service if DNS is unreachable without a loooong hang, but linux has getaddrinfo_a, so it should not be hard to implement it this way.

**Reply by fencepost-error** [reply]
Ah, I see, so you've got long-running connections and need to be able to find changed IP addresses without stalling things, that makes sense.

**Reply by fencepost-error** [reply]
Just wondering, have you had a chance to do anything with this?  I noticed this morning that MQTT data had stopped updating, what had happened is that one of the Modbus TCP devices had been assigned a new address via DHCP (I thought I'd fixed that but obviously not) and modmqttd exited.

By the looks of it it was because one of the three IP addresses it was trying to connect to was unreachable, in the meantime I'd spent a bit of time trying to figure out where the problem was so can't remember the original error message but the symptoms were that modmqttd had exited and trying to restart it produced error messages about being unable to connect.  Updating the config to match the new address fixed things.

**Reply by BlackZork** [reply]
> one of the Modbus TCP devices had been assigned a new address via DHCP (I thought I'd fixed that but obviously not) and modmqttd exited

Can you attach a log from this event? It looks like a bug. modmqttd is supposed to retry to connect forever in this case.

> Just wondering, have you had a chance to do anything with this? 

I've decided to use getaddrinfo_a, but this is very low on my priority list. It is easy fixable on DHCP side. Missing std.map(), templates are better candidates to work on than this DNS issue for now.

**Reply by fencepost-error** [reply]
Sure, here's the syslog stuff matching 'mqtt', I've cut out the date at the start (which is always the same) and the machine name.  192.168.10.42 was the device that was unreachable.  My guess is the two 'connected' entries are .79 and .80, with .42 being unreachable.
```
13:14:22 modmqttd[63602]: [INFO ] modmqttd is starting
13:14:22 modmqttd[63602]: [INFO ] Added converter plugin std
13:14:22 modmqttd[63602]: [INFO ] Added converter plugin expr
13:14:22 modmqttd[63602]: [INFO ] Connecting to 192.168.10.79:502
13:14:22 modmqttd[63602]: [INFO ] Connecting to 192.168.10.80:502
13:14:22 modmqttd[63602]: [INFO ] Response timeout set to 500ms
13:14:22 modmqttd[63602]: [INFO ] Response timeout set to 500ms
13:14:22 modmqttd[63602]: [INFO ] modbus: connecting
13:14:22 modmqttd[63602]: [INFO ] modbus: connecting
13:14:22 modmqttd[63602]: [INFO ] Connecting to 192.168.10.42:502
13:14:22 modmqttd[63602]: [INFO ] Response timeout set to 500ms
13:14:22 modmqttd[63602]: [INFO ] modbus: connecting
13:14:22 modmqttd[63602]: [INFO ] modbus: connected
13:14:22 modmqttd[63602]: [INFO ] Waiting for mqtt network to become online
13:14:22 modmqttd[63602]: [INFO ] modbus: connected
13:14:22 modmqttd[63602]: [INFO ] Waiting for mqtt network to become online
13:14:23 modmqttd[63602]: [INFO ] Waiting for mqtt network to become online
13:14:23 modmqttd[63602]: [INFO ] Waiting for mqtt network to become online
13:14:23 modmqttd[63602]: [INFO ] Connecting to localhost:1883
13:14:23 modmqttd[63602]: [INFO ] Connection established
13:14:23 modmqttd[63602]: [INFO ] Mqtt connected, sending subscriptions…
13:14:23 modmqttd[63602]: [INFO ] Mqtt ready to process messages
13:14:23 modmqttd[63602]: [ERROR] modbus connection failed(115) : Operation now in progress
13:14:23 modmqttd[63602]: [INFO ] modbus: connecting
13:14:23 modmqttd[63602]: [INFO ] Initial poll done in 565ms
13:14:23 modmqttd[63602]: [ERROR] modbus connection failed(115) : Operation now in progress
13:14:24 modmqttd[63602]: [INFO ] Initial poll done in 1809ms
13:14:33 modmqttd[63602]: [INFO ] modbus: connecting
13:14:34 modmqttd[63602]: [ERROR] modbus connection failed(115) : Operation now in progress
13:14:49 modmqttd[63602]: [INFO ] modbus: connecting
13:14:49 modmqttd[63602]: [ERROR] modbus connection failed(115) : Operation now in progress
13:15:09 modmqttd[63602]: [INFO ] modbus: connecting
13:15:10 modmqttd[63602]: [ERROR] modbus connection failed(115) : Operation now in progress
13:15:35 modmqttd[63602]: [INFO ] modbus: connecting
13:15:35 modmqttd[63602]: [ERROR] modbus connection failed(115) : Operation now in progress
13:16:05 modmqttd[63602]: [INFO ] modbus: connecting
13:16:06 modmqttd[63602]: [ERROR] modbus connection failed(115) : Operation now in progress
13:16:41 modmqttd[63602]: [INFO ] modbus: connecting
13:16:41 modmqttd[63602]: [ERROR] modbus connection failed(115) : Operation now in progress
13:16:45 modmqttd[63602]: [INFO ] Got SIGTERM, exiting…
13:16:45 modmqttd[63602]: [INFO ] Stopping modbus clients
13:16:45 modmqttd[63602]: [INFO ] Publishing availability status 0 for all registers
13:16:45 modmqttd[63602]: [INFO ] Disconnecting from mqtt broker
13:16:45 modmqttd[63602]: [INFO ] Disconnected from mqtt broker, code:No error.
13:16:45 modmqttd[63602]: [INFO ] Stopping mosquitto message loop
13:16:45 modmqttd[63602]: [INFO ] Shutdown finished
13:16:45 modmqttd[63602]: [INFO ] modmqttd stopped
13:16:45 systemd[1]: modmqttd.service: Deactivated successfully.
```

**Reply by BlackZork** [reply]
Thanks. I added IP address/device name to connection failed log message.
Nothing suspicious here, modmqttd received SIGTERM and exited normally.

**Reply by fencepost-error** [reply]
Yeah, the logs are a bit odd, I post the data to an external site which showed a last reading (before it stopped updating) at around 1am on 25 June, and restarted at around 2pm on 26 June, it reports in half-hour intervals so the ~2pm restart of readings would correspond to restarting modmqttd with the IP address corrected around 1:30pm or so.

The log entries from before it failed are:
```
13:58:52 modmqttd[180337]: [ERROR] Cannot execute any command in last 26s, reconnecting
13:58:52 modmqttd[180337]: [INFO ] modbus: connecting
13:58:52 modmqttd[180337]: [INFO ] modbus: connected
13:58:52 modmqttd[180337]: [INFO ] Initial poll done in 584ms
14:06:23 modmqttd[180337]: [ERROR] 1 error(s) when reading register 11.0, last error: libmodbus: read fn 0 failed with return code -1: Connection timed out
14:32:57 modmqttd[180337]: [ERROR] 1 error(s) when reading register 20.4, last error: libmodbus: read fn 4 failed with return code -1: Connection timed out
14:40:56 modmqttd[180337]: [ERROR] 1 error(s) when reading register 11.0, last error: libmodbus: read fn 0 failed with return code -1: Connection timed out
14:44:29 modmqttd[180337]: [ERROR] 1 error(s) when reading register 20.5, last error: libmodbus: read fn 5 failed with return code -1: Connection timed out
16:20:08 modmqttd[180337]: [ERROR] 1 error(s) when reading register 20.5, last error: libmodbus: read fn 5 failed with return code -1: Connection timed out
16:32:35 modmqttd[180337]: [ERROR] 1 error(s) when reading register 12.1, last error: libmodbus: read fn 1 failed with return code -1: Connection timed out
18:56:42 modmqttd[180337]: [ERROR] 1 error(s) when reading register 10.0, last error: libmodbus: read fn 0 failed with return code -1: Connection timed out
18:56:42 modmqttd[180337]: [ERROR] Cannot execute any command in last 20s, reconnecting
18:56:42 modmqttd[180337]: [INFO ] modbus: connecting
18:56:42 modmqttd[180337]: [INFO ] modbus: connected
18:56:42 modmqttd[180337]: [INFO ] Initial poll done in 580ms
18:59:10 modmqttd[180337]: [ERROR] 1 error(s) when reading register 11.264, last error: libmodbus: read fn 264 failed with return code -1: Connection timed out
19:25:45 modmqttd[180337]: [ERROR] 1 error(s) when reading register 11.0, last error: libmodbus: read fn 0 failed with return code -1: Connection timed out
```
These are all for 24 June.  So the last entry reporting a problem was 7:25pm on 24 June, the external site stopped updating at around 1am on 25 June, and then I fixed things and restarted around 2pm on 26 June.  I also used MQTT Explorer before then to see whether the MQTT data was being updated and it wasn't, the values were static.

The SIGTERM was me doing a 'systemctl restart modmqttd' after updating the IP address for the device, sorry, nothing to do with any errors.  So the timeline was:

24 June 19:25: Last syslog error entry from modmqttd.
25 June 01:00 (approx): MQTT values stop being updated.
26 June 13:16 (approx): Restart modmqttd with corrected IP address.
26 June 14:00 (approx): MQTT values start being updated again.

So it looks like there was some sort of silent failure around 1am on 25 June, which was when the IP address was reassigned by the DHCP server.

**Reply by fencepost-error** [reply]
In the meantime I've added a workaround to my script compiler that swaps out DNS names for IP addresses:
```
# Create the header.  This performs DNS lookups on the named devices and
# inserts the corresponding IP addresses into the header for the config
# file

compile_header()
	{
	for DNSNAME in "$@" ; do

		# Get the IP address of the device.  This isn't very fancy, it just
		# returns the first one found without much checking on the assumption
		# that an embedded modbus gateway isn't going to be multihomed.
		IP_ADDRESS=$(dig +short "$DNSNAME" | head -n1)
		if [ -z "$IP_ADDRESS" ] ; then
			echo "No DNS entry for server $DNSNAME" >&2 ;
			exit 1 ;
		fi

		if [ $VERBOSE -eq 1 ] ; then
			echo "Setting address for ${DNSNAME} to ${IP_ADDRESS}" ;
		fi

		# Swap out the DNS name for the IP address
		sed -i s/"${DNSNAME}"/"${IP_ADDRESS}"/g ${OUTPUT_FILE}

	done
	}
```
I really need to set up a repo for this so others can use it too...
